Enhance SlurmClusterResolver

Flamefire opened the pull request:
This enhances the SlurmClusterResolver and fixes its bugs, see #36094

The design goal is that the user simply creates an instance of the class when the python script is run by `srun` (the `mpirun` wrapper on SLURM). As all information is available in the environment this is certainly possible.

- Default all arguments to something reasonable
- Use Slurm step env variables to fill in data

This has been tested (kinda) successfully on a SLURM cluster with TF 2.1.0 and MultiWorkerMirroredStrategy. It works with artificial data. I couldn't test it fully with MNIST due to https://github.com/tensorflow/tensorflow/issues/36153

What should be discussed besides the code:
- Usage of `nvidia-smi` to query the number of GPUs. I couldn't use `context.num_gpus()` at this point because this initializes the context including collectives which is not wanted as collectives might be overwritten later when creating a strategy

Flamefire committed:
Enhance SlurmClusterResolver

Default all arguments to something reasonable
Use Slurm step env variables to fill in data

frankchn commented on tensorflow/python/distribute/cluster_resolver/slurm_cluster_resolver.py:
```diff
@@ -27,6 +27,44 @@
 from tensorflow.python.training.server_lib import ClusterSpec
 from tensorflow.python.util.tf_export import tf_export
 
+from hostlist import expand_hostlist, parse_slurm_tasks_per_node
```
Can we reimplement `expand_hostlist` and `parse_slurm_tasks_per_node` here? Given the size of TensorFlow as it is, we are trying not to include new packages unless absolutely essential.

I note that `hostlist` is GPL so it is not compatible with TensorFlow's Apache license.

Flamefire replied:
Yes could be done, but the format is... special. I guess even though I would not (and did not) look at the source code of hostlist I assume I'll end up with an implementation VERY similar. This could be a license issue too.

I wasn't aware that consuming a GPL Python package from another Apache Python package is not allowed. I mean: It doesn't ship with it. The users installs it.

frankchn replied:
I am not a lawyer, and I am guessing it depends on the interpretation of what "linking" is under GPL, but if calling a function is considered linking, then we may want to avoid this if possible.

https://www.apache.org/licenses/GPL-compatibility.html

frankchn commented on tensorflow/python/distribute/cluster_resolver/slurm_cluster_resolver.py:
```diff
@@ -27,6 +27,44 @@
+  try:
+    return os.environ[name]
+  except KeyError:
+    raise RuntimeError('%s not found in environment. Not running inside a SLURM step?' % name)
```
Python max line length is 80.

Flamefire replied:
IIRC Python line length is even 79. Is there some style file or formatter for TF?

Flamefire replied:
E.g. the existing imports are already 93 chars long

frankchn replied:
Imports are excluded from the line-length requirement. Here is the standard Google Python style guide, which we follow: http://google.github.io/styleguide/pyguide.html and there are some links to a formatter and vim settings file.

frankchn commented on tensorflow/python/distribute/cluster_resolver/slurm_cluster_resolver.py:
```diff
@@ -95,20 +126,33 @@ def __init__(self,
+      jobs = {'worker': num_tasks}
+    if gpus_per_node is None:
+      gpus_per_node = sum(l.startswith('GPU ') for l in
+                          subprocess.check_output(['nvidia-smi', '--list-gpus']).
```
Can we abstract this out to another function? The function should also throw an exception if `nvidia-smi` is not found for instance.

Flamefire replied:
Sure

frankchn reviewed (changes requested):
Thanks for doing this -- the general approach looks reasonable. I am also adding @steven-chien, who wrote the original `SlurmClusterResolver` for comments.

frankchn commented:
cc @steven-chien

steven-chien commented:
Thank you very much for the work @Flamefire and apologies for the late reply. I made the resolver originally for a work that needs a simple pattern of one/two GPUs per process so support for advanced resource distribution is not implemented. I always wanted to extend it but then other works came along...

With the complete shift to Slurm env variables, one problem that I see is things might behave differently when (different implementation of) mpirun is used instead of srun, i.e. OMPI_COMM_WORLD_SIZE is used instead, etc, so this also should be tested.

For expand_hostlist:
- My original plan was to use expand_hostlist as well but eventually skipped it due to license and dependency issues. Thus the scontrol solution that you saw.

For auto query of GPUs:
- The auto query of the number of GPUs on a node is a very nice feature. However, the use of nvidia-smi reports all the GPUs used in a node.
- One problem with my original implementation is that the exclusive use of a node is assumed. In case of non-exclusive access of node, SLURM uses the environment variable CUDA_VISIBLE_DEVICES to assign GPUs, whereas nvidia-smi reports all GPUs present on a system.
- Some thought should be given on how to figure out how to take this into account in the future. (i.e. extract info from more Slurm variables)

Apart from that, I don't see any major issues with the code. Since Slurm is mostly used in HPC community I also suggest that we can improve the documentation, so people who wants to use TF on HPC clusters can use this easily.

Flamefire commented:
> With the complete shift to Slurm env variables, one problem that I see is things might behave differently when (different implementation of) mpirun is used instead of srun, i.e. OMPI_COMM_WORLD_SIZE is used instead, etc, so this also should be tested.

IIRC when SLURM is used then you'd better use srun too which also takes care of job steps. So I'd remove OMPI variable references to avoid people thinking this is for MPI and focus on correct work with SLURM. If someone wants/needs MPI support, they can implement another class or derive from this or use the ctor parameters.

>  My original plan was to use expand_hostlist as well but eventually skipped it due to license and dependency issues.

Ok so that needs to be rewritten. I'll look into that. scontrol is not really an option as it may not handle job steps correctly. But I'll check that first. Job step handling is crucial!

> However, the use of nvidia-smi reports all the GPUs used in a node.

That's what it is used for: To set gpus_per_node

> One problem with my original implementation is that the exclusive use of a node is assumed. In case of non-exclusive access of node, SLURM uses the environment variable CUDA_VISIBLE_DEVICES to assign GPUs,

That's actually a good idea: If CUDA_VISIBLE_DEVICES is set, then that is used instead of the nvidia-smi solution. This can (and should) be mentioned as a caveat though.

> Some thought should be given on how to figure out how to take this into account in the future. (i.e. extract info from more Slurm variables)

I did. There are no standard SLURM variables to use. Hence nvidia-smi to figure out the number of GPUs/node is the only thing you got. Defaulting the gpus_per_task to gpus_per_node//tasks_per_node is the most reasonable thing to do here.

> Since Slurm is mostly used in HPC community I also suggest that we can improve the documentation, so people who wants to use TF on HPC clusters can use this easily.

Yes this is what I intend: Currently TF on HPC is hard to awkward to use. Figuring out what to set how is difficult. And then try to explain that to users who just want to "make it work". The default arguments allow exactly this use case. I assume 90% of the situations are covered by that and then you can always use the ctor args to overwrite values that are gotten wrong or are non-standard.

steven-chien commented:
> > With the complete shift to Slurm env variables, one problem that I see is things might behave differently when (different implementation of) mpirun is used instead of srun, i.e. OMPI_COMM_WORLD_SIZE is used instead, etc, so this also should be tested.
> 
> IIRC when SLURM is used then you'd better use srun too which also takes care of job steps. So I'd remove OMPI variable references to avoid people thinking this is for MPI and focus on correct work with SLURM. If someone wants/needs MPI support, they can implement another class or derive from this or use the ctor parameters.
> 
This one is tricky on generality. As some centers use neither srun or mpirun, but some other wrappers to optimize node to process mapping. We might initially try to explicitly document that a user should only use this Slurm with srun that is built with PMI.

> > Some thought should be given on how to figure out how to take this into account in the future. (i.e. extract info from more Slurm variables)
> 
> I did. There are no standard SLURM variables to use. Hence nvidia-smi to figure out the number of GPUs/node is the only thing you got. Defaulting the gpus_per_task to gpus_per_node//tasks_per_node is the most reasonable thing to do here.
> 
One thing to be careful here is that if the user passes --gres during srun instead of salloc, the environment variable might be overwritte. see GPU management in https://slurm.schedmd.com/gres.html
 `This environment variable is only set when tasks are launched on a specific compute node (no global environment variable is set for the salloc command and the environment variable set for the sbatch command only reflects the GPUs allocated to that job on that node, node zero of the allocation).`

Flamefire commented:
Ok. I think we can document the environment variables used so when users are not using standard SLURM tools, they can check if they still are able to use this. Maybe someone from such a center could help with compatibility.

I'm purposely writing this for execution inside `srun`. Only then is the final configuration known. So salloc/sbatch settings are not covered here.

frankchn commented:
Thanks for your comments @steven-chien! It would be good to have documentation to cover standard Slurm use cases for this written.

Flamefire commented:
What is the preferred way of writing test assertions? pytest style `assert foo == bar` or unittest style `self.assertEqual(foo, bar)`? I couldn't find this in the docu and at least this file seems to be inconsistent. I'd prefer the pytest style for improved clarity

frankchn commented:
@Flamefire The current preferred usage is in the `unittest` style with `assertEqual`, `assertNotEqual`, etc...

Flamefire committed:
Remove dependency on python-hostlist and fix tests

Also refactor some functions and improve the documentation

Flamefire commented:
I rebased to master and addressed the comments:
- Fix line lengths to 80 (as far as I could find, is there a style check for TF?)
- Implement `expand_hostlist` and `expand_tasks_per_node` to drop dependency on python-hostlist
- Improve some error reporting
- Prefer `CUDA_VISIBLE_DEVICES` over `nvidia-smi`
- Add `_resolve_*` as member functions so they can be overwritten for e.g. MPI environments
- Only keep  functions in the public "namespace" that might be useful for others
- Improve documentation (I'll probably have to update `Readme.slurm` too but I'd prefer to rename it to `Slurm.md` or `ReadmeSlurm.md` to get syntax support from Github)
- Fix existing tests and add own ones for the hostlist stuff and a default constructed SlurmClusterResolver

frankchn commented on tensorflow/python/distribute/cluster_resolver/slurm_cluster_resolver.py:
```diff
@@ -21,94 +21,236 @@
+      else:
+        hosts.extend(prefix + str(i)
+                     for i in expand_range_expression(m.group(3))
+        )
```
Just leave this on the previous line with the other closing brackets

frankchn commented on tensorflow/python/distribute/cluster_resolver/slurm_cluster_resolver_test.py:
```diff
@@ -29,8 +29,23 @@
-  def mock_resolve_hostnames_output(self):
-    return ['t02n13', 't02n41', 't02n43', 't02n44']
+  def test_expand_hostlist(self):
+    from tensorflow.python.distribute.cluster_resolver.slurm_cluster_resolver import expand_hostlist
```
Just add the import at the top.

frankchn commented on tensorflow/python/distribute/cluster_resolver/slurm_cluster_resolver_test.py:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+# Copyright 2018-2020 The TensorFlow Authors. All Rights Reserved.
```
Change to `Copyright 2020 The TensorFlow Authors.`

frankchn commented on tensorflow/python/distribute/cluster_resolver/slurm_cluster_resolver_test.py:
```diff
@@ -29,8 +29,23 @@
+                     ['n1', 'n2', 'm5', 'o3', 'o4', 'o6', 'o7', 'o8', 'o9'])
+
+  def test_expand_tasks_per_node(self):
+    from tensorflow.python.distribute.cluster_resolver.slurm_cluster_resolver import expand_tasks_per_node
```
as above.

frankchn commented:
Thanks for doing this @Flamefire! I've added some minor comments, and we should be good to go once those are fixed. If the internal linter catches anything else, I'll fix it then.

Flamefire committed:
Adress review comments

Flamefire commented:
Done. For completeness: There is `scontrol show hostnames=n[1-3]` which does the hostlist expansion. For an invalid hostlist it shows "Invalid hostlist: <input>" but exits with no error code. It would significantly reduce the code as `expand_hostlist` would become a single call and split of output potentially with a check for "invalid hostlist".   
Downside: It is harder to test as it'd need to be mocked and I don't know how long this was available and if the "invalid" output changed or is the same for all versions. The current implementations is more independent.

frankchn commented:
@Flamefire You may have to run the API compatibility test/updater at https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/api/tests/api_compatibility_test.py#L23

Flamefire commented:
Ah yes, of course the defaults have changed to something more reasonable:

```
    member_method {
      name: "__init__"
-     argspec: "args=[\'self\', \'jobs\', \'port_base\', \'gpus_per_node\', \'gpus_per_task\', \'tasks_per_node\', \'auto_set_gpu\', \'rpc_layer\'], varargs=None, keywords=None, defaults=[\'8888\', \'1\', \'1\', \'None\', \'True\', \'grpc\'], "
?                                                                                                                                                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+     argspec: "args=[\'self\', \'jobs\', \'port_base\', \'gpus_per_node\', \'gpus_per_task\', \'tasks_per_node\', \'auto_set_gpu\', \'rpc_layer\'], varargs=None, keywords=None, defaults=[\'None\', \'8888\', \'None\', \'None\', \'None\', \'True\', \'grpc\'], "
?                                                                                                                                                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Can I just copy the new defaults to tensorflow/tools/api/golden/v2/tensorflow.distribute.cluster_resolver.-slurm-cluster-resolver.pbtxt? I guess that would make it pass but: 
> You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

Flamefire committed:
Update golden API for slurm-cluster-resolver

Flamefire committed:
Fix style issues

Flamefire commented:
@frankchn This is currently still not easily usable due to https://github.com/tensorflow/tensorflow/issues/36550

I have some code locally for that:
```
def create_tf_config(num_worker=None, num_ps=0, trial=1, job=None, **kwargs):
    """Create a TensorFlow config suitable for TF_CONFIG
    Args:
      num_worker: Number of worker tasks, defaults to available tasks in current SLURM job (step)
      num_ps: Number of parameter server tasks
      trial: The identifier of the hyperparameter tuning trial currently running.
      job: The job parameters you used when you initiated the job.
    """

    if num_worker is None:
        num_worker = get_num_slurm_tasks() - num_ps
    assert num_worker > 0
    assert num_ps >= 0

    jobs = {'worker': num_worker}
    if num_ps:
        jobs['ps'] = num_ps
    resolver_args = {'jobs': jobs}
    resolver_args.update(kwargs)
    resolver = SlurmClusterResolver(**resolver_args)
    return {
        'cluster': resolver.cluster_spec().as_dict(),
        'task': {
            'type': resolver.get_task_info()[0],
            'index': resolver.get_task_info()[1],
            'trial': trial
        },
        'rpc_layer': resolver.rpc_layer,
        'job': job
    }
```

I think it would make sense to set `TF_CONFIG` in the cluster resolver, at least until #36550 is resolved. Or provide some function for setting the TF_CONFIG variable from a cluster resolver (`return` part of the code). What do you think? How and where would you suggest to add this?

frankchn commented:
@Flamefire Thanks for filing the bug! I would prefer the current changes you have go in first (it is undergoing internal review now) and let's see how #36550 shakes out since I am not that familiar with the current usages there.

Flamefire commented:
@frankchn What about the documentation update? https://github.com/tensorflow/tensorflow/blob/25d2d106cbe84969809fe546507a09c2e7f653f9/tensorflow/python/distribute/cluster_resolver/README.slurm is out of date now. I can update that with the description of "my" SlurmClusterResolver and how to use.

Related: I recently wrote an MPIClusterResolver inheriting from SlurmClusterResolver which uses mpi4py (i believe already a transitive dependency of TF) to fill in the template methods created by this. I can contribute that too if wanted.

frankchn commented:
@Flamefire Sure, both of those will be super helpful! Thanks for contributing to TensorFlow :)

frankchn commented:
Just create PRs and assign me as the reviewer.

Flamefire commented:
Ok, will do. Is there some auto-formatter that can be used? autopep8? yapf? Haven't seen a style definition file in the repo.

frankchn commented:
According to the [Google Python style guide](http://google.github.io/styleguide/pyguide.html), yapf with `--style google` should work.

frankchn commented:
The indentation is a major difference between our internal style guide and the public one, so you should use 2 spaces.
